controllers/capabilities/tenant_controller.go

```diff
@@ -95,119 +107,116 @@ func (r *TenantReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
-		return ctrl.Result{}, nil
-	}
+	reqLogger.Info("Tenant reconciled successfully")
+	return ctrl.Result{}, nil
+}
 
-	// If the tenant.Status.TenantID is found and the annotation is not found - create
-	// If the tenant.Status.TenantID is found and the annotation is found but, the value of annotation is different to the status.TenantID - update
-	tenantIdAnnotationFound := true
-	tenantId := tenantR.Status.TenantId
-	if value, found := tenantR.ObjectMeta.Annotations[tenantIdAnnotation]; tenantId != 0 && !found || tenantId != 0 && found && value != strconv.FormatInt(tenantR.Status.TenantId, 10) {
-		if tenantR.ObjectMeta.Annotations == nil {
-			tenantR.ObjectMeta.Annotations = make(map[string]string)
-		}
+func (r *TenantReconciler) specReconciler(tenantCR *capabilitiesv1alpha1.Tenant, logger logr.Logger) (*TenantStatusReconciler, error) {
+	// Save a copy of the original tenant CR prior to changing it for status reconciler
+	originalTenantCR := tenantCR.DeepCopy()
 
-		tenantR.ObjectMeta.Annotations[tenantIdAnnotation] = strconv.FormatInt(tenantR.Status.TenantId, 10)
-		tenantIdAnnotationFound = false
+	// Setup porta client
+	portaClient, err := r.setupPortaClient(tenantCR, originalTenantCR, logger)
+	if err != nil {
+		statusReconciler := NewTenantStatusReconciler(r.BaseReconciler, tenantCR, originalTenantCR, err)
+		return statusReconciler, err
 	}
 
-	tenantFinalizerFound := true
-	if !controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
-		controllerutil.AddFinalizer(tenantR, tenantFinalizer)
-		tenantFinalizerFound = false
+	// Process tenant deletion
+	tenantInDeletion, err := r.reconcileDeletion(tenantCR, portaClient, logger)
+	if err != nil || tenantInDeletion {
+		statusReconciler := NewTenantStatusReconciler(r.BaseReconciler, tenantCR, originalTenantCR, err)
+		return statusReconciler, err
 	}
 
-	if !tenantFinalizerFound || !tenantIdAnnotationFound {
-		err = r.UpdateResource(tenantR)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		return ctrl.Result{}, nil
-	}
+	// Adding annotations and finalizers
+	changedMetadata := r.reconcileMetadata(tenantCR)
 
-	changed := tenantR.SetDefaults()
-	if changed {
-		err = r.UpdateResource(tenantR)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		reqLogger.Info("Tenant resource updated with defaults")
-		// Expect for re-trigger
-		return ctrl.Result{}, nil
+	// Bring tenant secret ref and namespace to lower if setup incorrectly
+	changedDefaults := tenantCR.SetDefaults()
+
+	if changedMetadata || changedDefaults {
+		statusReconciler := NewTenantStatusReconciler(r.BaseReconciler, tenantCR, originalTenantCR, err)
+		return statusReconciler, err
 	}
 
-	internalReconciler := NewTenantInternalReconciler(r.BaseReconciler, tenantR, portaClient, reqLogger)
-	res, err := internalReconciler.Run()
+	// Main logic for reconciling the tenant CR against 3scale
+	internalReconciler := NewTenantInternalReconciler(r.BaseReconciler, tenantCR, portaClient, logger)
```

Comment:
I would rename it to `TenantThreescaleReconciler`

no need for a "internal reconciler"

Reply:
Updated, thank you